### PR TITLE
Precompile events allow list

### PR DESCRIPTION
**Pull request by abi87**
## Why this should be merged

Adds event to allow list functions. 

## How this works

Adds new Event to ABI and regenerated with the precompile generation tool.

`event RoleSet(uint256 indexed role, address indexed account, address indexed sender, uint256 oldRole);`

Increases total gas cost by 2131 (base event gas, 4 topics gas, 32-byte unindexed data)

## How this was tested

Added UT and hardhat tests.

## How is this documented

Will be documented with other Durango changes

### CODE REVIEW

File: contracts/contracts/interfaces/IAllowList.sol [hunk 1]
```diff
@@ -2,6 +2,11 @@
 pragma solidity ^0.8.0;
 
 interface IAllowList {
+  event AdminAdded(address indexed sender, address indexed account);
+  event EnabledAdded(address indexed sender, address indexed account);
+  event ManagerAdded(address indexed sender, address indexed account);
+  event RoleRemoved(address indexed sender, address indexed account);
```
Comment (aaronbuchwald) [inline contracts/contracts/interfaces/IAllowList.sol:8]: I think the event should include this information (and in this order of priority with respect to whether it's indexed).

1. modified account
2. new role
3. account that caused the modification
4. previous role

Since that's 4 fields, one of them would need to be captured in the function name or we'd need to make at least one non-indexed.

I'd lean towards making it a single event and leaving the previous role un-indexed since it's unlikely imo a user would need to search all of the allow list modifications that changed the status from manager -> X.

File: precompile/allowlist/event.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+		return PackAdminAddedEvent(sender, account)
+	case ManagerRole:
+		return PackManagerAddedEvent(sender, account)
+	case EnabledRole:
+		return PackEnabledAddedEvent(sender, account)
+	case NoRole:
+		return PackRoleRemovedEvent(sender, account)
+	default:
+		panic("invalid role")
+	}
```
Comment (aaronbuchwald) [inline precompile/allowlist/event.go:52]: can we return an error instead of panic-ing here? I'd prefer defense in depth from something that should never happen to a panic that will shut down the whole chain.

File: precompile/allowlist/test_allowlist.go [hunk 3]
```diff
@@ -539,6 +539,191 @@ func AllowListTests(t testing.TB, module modules.Module) map[string]testutils.Pr
 				require.Equal(t, EnabledRole, GetAllowListStatus(state, contractAddress, TestNoRoleAddr))
 			},
 		},
+		"allowList admin event doesn't log if D fork is not active": {
```
Comment (aaronbuchwald) [inline precompile/allowlist/test_allowlist.go:542]: Can we adjust these names?

I think a better naming convention is set above with the address of the caller, the role being set, and then adding pre-DUpgrade

For example, this one could be "admin set admin pre-DUpgrade"

File: contracts/test/contract_deployer_allow_list.ts [hunk 4]
```diff
@@ -48,3 +53,40 @@ describe("ExampleDeployerList", function () {
...
+    await expect(contract.setAdmin(testAddress))
+      .to.emit(contract, 'AdminAdded')
+      .withArgs(owner.address, testAddress)
+  })
+
+  it("should emit manager address added event", async function () {
+    let testAddress = "0x0222000000000000000000000000000000000002"
+    await expect(contract.setManager(testAddress))
+      .to.emit(contract, 'ManagerAdded')
+      .withArgs(owner.address, testAddress)
```
Comment (aaronbuchwald) [inline contracts/test/contract_deployer_allow_list.ts:76]: These assertions are sick ❤️ , so easy

**Review by aaronbuchwald — CHANGES_REQUESTED** [review]

File: precompile/allowlist/test_allowlist.go (hunk 3 above)
Comment (ceyonur) [reply precompile/allowlist/test_allowlist.go]: done

File: contracts/contracts/interfaces/IAllowList.sol (hunk 1 above)
Comment (ceyonur) [reply contracts/contracts/interfaces/IAllowList.sol]: done

File: contracts/test/contract_deployer_allow_list.ts [hunk 5]
```diff
@@ -2,7 +2,12 @@
 // See the file LICENSE for licensing terms.
 
 import { ethers } from "hardhat"
-import { test } from "./utils"
+import { Roles, test } from "./utils"
+import { SignerWithAddress } from "@nomiclabs/hardhat-ethers/signers";
+import { expect } from "chai";
+import {
+  Contract,
+} from "ethers"
```
Comment (darioush) [inline contracts/test/contract_deployer_allow_list.ts:10]:
```suggestion
import { Contract } from "ethers"
```

File: contracts/test/tx_allow_list.ts [hunk 6]
```diff
@@ -2,7 +2,12 @@
 // See the file LICENSE for licensing terms.
 
 import { ethers } from "hardhat"
-import { test } from "./utils"
+import { Roles, test } from "./utils"
+import { SignerWithAddress } from "@nomiclabs/hardhat-ethers/signers";
+import { expect } from "chai";
+import {
+  Contract,
+} from "ethers"
```
Comment (darioush) [inline contracts/test/tx_allow_list.ts:10]:
```suggestion
import { Contract } from "ethers"
```

File: precompile/allowlist/role.go [hunk 7]
```diff
@@ -77,18 +80,18 @@ func (r Role) Hash() common.Hash {
...
+		return "setManager", nil
 	case EnabledRole:
-		return "setEnabled"
+		return "setEnabled", nil
 	case NoRole:
-		return "setNone"
+		return "setNone", nil
 	default:
-		panic("unknown role")
+		return "", ErrInvalidRole
```
Comment (darioush) [inline precompile/allowlist/role.go:94]: seems this err is ignored if it is returned?

File: precompile/allowlist/test_allowlist.go [hunk 8]
```diff
@@ -276,16 +324,28 @@ func AllowListTests(t testing.TB, module modules.Module) map[string]testutils.Pr
...
+				require.Len(t, logsTopics, 1)
+				require.Len(t, logsData, 1)
+				topics := logsTopics[0]
+				require.Len(t, topics, 4)
+				require.Equal(t, AllowListABI.Events["RoleSet"].ID, topics[0])
+				require.Equal(t, NoRole.Hash(), topics[1])
+				require.Equal(t, TestNoRoleAddr.Hash(), topics[2])
+				require.Equal(t, TestManagerAddr.Hash(), topics[3])
+				data := logsData[0]
+				require.Equal(t, NoRole.Bytes(), data)
```
Comment (darioush) [inline precompile/allowlist/test_allowlist.go:345]: could we abstract this to a helper function?

File: precompile/allowlist/role.go (hunk 7 above)
Comment (ceyonur) [reply precompile/allowlist/role.go:94]: we handle the err in case we call it with an arbitrary role. I.e calls like role.GetSetterFunctionName without a defined Role type Admin/Enabled etc.

**Review by darioush — DISMISSED** [review]

File: contracts/contracts/interfaces/IAllowList.sol [hunk 9]
```diff
@@ -2,10 +2,7 @@
 pragma solidity ^0.8.0;
 
 interface IAllowList {
-  event AdminAdded(address indexed sender, address indexed account);
-  event EnabledAdded(address indexed sender, address indexed account);
-  event ManagerAdded(address indexed sender, address indexed account);
-  event RoleRemoved(address indexed sender, address indexed account);
+  event RoleSet(uint256 indexed role, address indexed account, address indexed sender, uint256 oldRole);
```
[the remark this answers is not in the document]
Comment (aaronbuchwald) [reply contracts/contracts/interfaces/IAllowList.sol:5]: We should update the name for this throughout imo

File: precompile/allowlist/event.go [hunk 10]
```diff
@@ -5,49 +5,33 @@
...
-		return PackEnabledAddedEvent(sender, account)
-	case NoRole:
-		return PackRoleRemovedEvent(sender, account)
-	default:
-		panic("invalid role")
+// UnpackRoleSetEventData attempts to unpack non-indexed [dataBytes].
+func UnpackRoleSetEventData(dataBytes []byte) (Role, error) {
+	eventData := struct {
+		OldRole *big.Int
+	}{}
```
Comment (aaronbuchwald) [inline precompile/allowlist/event.go:31]: Is this not auto-generated by the precompile template?

File: precompile/allowlist/allowlist.go [hunk 11]
```diff
@@ -181,11 +187,24 @@ func CreateAllowListPrecompile(precompileAddr common.Address) contract.StatefulP
...
+		} else if noRoleFnName, _ := NoRole.GetSetterFunctionName(); name == noRoleFnName {
+			fn = contract.NewStatefulPrecompileFunction(method.ID, createAllowListRoleSetter(precompileAddr, NoRole))
+		} else if managerFnName, _ := ManagerRole.GetSetterFunctionName(); name == managerFnName {
+			fn = contract.NewStatefulPrecompileFunctionWithActivator(method.ID, createAllowListRoleSetter(precompileAddr, ManagerRole), contract.IsDUpgradeActivated)
+		} else {
+			panic(fmt.Sprintf("unexpected method name: %s", name))
+		}
+		functions = append(functions, fn)
+	}
+	return functions
```
Comment (aaronbuchwald) [inline precompile/allowlist/allowlist.go:209]: what's the reason for this change?

**Review by aaronbuchwald — CHANGES_REQUESTED** [review]

File: precompile/allowlist/event.go (hunk 10 above)
Comment (ceyonur) [reply precompile/allowlist/event.go:31]: it was generated but IMO it's better UX to convert these inline to use higher-level structs (like roles etc).

File: precompile/allowlist/allowlist.go (hunk 11 above)
Comment (ceyonur) [reply precompile/allowlist/allowlist.go:209]: We introduced ABI to allow list. We also return error from `GetSetterFunctionName`.

**Review by aaronbuchwald — APPROVED** [review]
